src/ui/GameUI.cpp

```diff
@@ -0,0 +1,358 @@
+            text->setText("22 / 100 XP");
+            text->setFont(smallfont);
+            text->setColor(0xff888888);
+            text->setHJustify(Field::justify_t::CENTER);
+            text->setVJustify(Field::justify_t::CENTER);
+        }
+    }
+
+    // hunger icon
+    auto hunger = frame->addImage(
+        SDL_Rect{16, frame->getSize().h - 160, 64, 64},
+        0xffffffff,
+        "images/system/Hunger.png",
+        "hunger"
+    );
+
+    // hp
+    {
+        Frame* hp = frame->addFrame("hp");
+        hp->setColor(0xffffffff);
+        hp->setSize(SDL_Rect{16, frame->getSize().h - 96, 40, 24});
+        hp->setActualSize(SDL_Rect{0, 0, 40, 24});
+        {
+            auto red = hp->addImage(
+                SDL_Rect{4, 4, 32, 16},
+                0xff8888ff,
+                "images/system/white.png",
+                "red"
+            );
+        }
+        Frame* hp_bar = frame->addFrame("hp_bar");
+        hp_bar->setColor(0xffffffff);
+        hp_bar->setSize(SDL_Rect{56, frame->getSize().h - 96, 256, 24});
+        hp_bar->setActualSize(SDL_Rect{0, 0, 256, 24});
+        {
+            auto red = hp_bar->addImage(
+                SDL_Rect{4, 4, hp_bar->getSize().w - 8, 16},
+                0xff8888ff,
+                "images/system/white.png",
+                "red"
+            );
+            Field* text = hp_bar->addField("text", 64);
+            text->setSize(SDL_Rect{6, 3, 64, hp_bar->getSize().h - 4});
+            text->setText("HP");
+            text->setFont(smallfont);
+            text->setColor(0xffffffff);
+            text->setHJustify(Field::justify_t::LEFT);
+            text->setVJustify(Field::justify_t::CENTER);
+        }
+    }
+
+    // mp
+    {
+        Frame* mp = frame->addFrame("mp");
+        mp->setColor(0xffffffff);
+        mp->setSize(SDL_Rect{16, frame->getSize().h - 68, 40, 24});
+        mp->setActualSize(SDL_Rect{0, 0, 40, 24});
+        {
+            auto blue = mp->addImage(
+                SDL_Rect{4, 4, 32, 16},
+                0xffff8888,
+                "images/system/white.png",
+                "blue"
+            );
+        }
+        Frame* mp_bar = frame->addFrame("mp_bar");
+        mp_bar->setColor(0xffffffff);
+        mp_bar->setSize(SDL_Rect{56, frame->getSize().h - 68, 64, 24});
+        mp_bar->setActualSize(SDL_Rect{0, 0, 64, 24});
+        {
+            auto blue = mp_bar->addImage(
+                SDL_Rect{4, 4, mp_bar->getSize().w - 8, 16},
+                0xffff8888,
+                "images/system/white.png",
+                "blue"
+            );
+            Field* text = mp_bar->addField("text", 64);
+            text->setSize(SDL_Rect{6, 3, 64, mp_bar->getSize().h - 4});
+            text->setText("MP");
+            text->setFont(smallfont);
+            text->setColor(0xffffffff);
+            text->setHJustify(Field::justify_t::LEFT);
+            text->setVJustify(Field::justify_t::CENTER);
```

Comment:
Man, I wish we could have templates for these MP and HP bars. I feel like half this file is just duplicating this code for the player, their allies, and other players.

Reply:
The intent eventually would be 2 image assets specifically for the player's HP/MP, then a different slimmer style for allies - ATM the wireframes are just rectangles so it's all very samey. Besides if we only write the code once here there's no point refactoring into a fancy catch-all draw function that'll eventually stop being suitable.

Reply:
I'm totally open to the idea that some code here could be consolidated but these are small and unique elements that aren't gonna be replicated any more so I'm on the same line as ben here